### PR TITLE
TW-2155(fix): normalize phonebook numbers to E.164 for lookup hashing

### PR DESCRIPTION
**Pull request by jbrnzl**
## Root cause / context
Phone numbers coming from local phonebooks can be stored in inconsistent formats (local/national formats, punctuation, spacing, extensions).  
During contact sync, hashing for Matrix lookup depended on those values being clean enough, but there was no strict normalization gate at ingestion time to guarantee a valid E.164 interpretation before MSISDN hashing. (ex: "0612345678" was valid to send for lookup as an msisdn. The same as "1234".)

## Solution
- Added phone_numbers_parser: to parse/validate raw phonebook numbers and normalize them to E.164 before hashing. (Why not flutter_libphonenumber ? Because it adds native initialization and heavier plumbing, while we only need lightweight batch E.164 normalization, which phone_numbers_parser already covers in pure Dart without startup overhead)
- Added sim_country_code_plus: to infer default country context (SIM first, locale fallback) for local numbers without +.
- Added iOS Info.plist key: NSCarrierUsageDescription: required by sim_country_code_plus to access carrier info and resolve SIM country code.
- Added strict phone normalization before creating domain `PhoneNumber` objects:
  - `tryNormalizePhoneNumberToE164(rawNumber, callerIsoCode)`
  - invalid/unresolvable numbers are dropped early to avoid surcharging server
- Added `SimCountryProvider` to provide country context for local numbers:
  - SIM country first
  - locale fallback (`fr_FR` and `fr-FR` like formats supported)
- Wired `SimCountryProvider` in DI and updated `PhonebookContactDatasourceImpl` constructor usage.
- Kept existing MSISDN sanitization in `PhoneNumber` (`msisdnSanitizer`) as a final guardrail before hash generation.
- Added/updated tests:
  - datasource tests now assert normalized E.164 output and filtering behavior
  - interactor test captures `lookupMxid` request addresses to confirm expected hashes are actually sent
  - dedicated normalizer unit tests for multiple real-world formats and edge cases

### Before this change:
local phone from address book
-> PhoneNumber(raw)
-> msisdnSanitizer
-> hash
-> lookup
### After this change:
local phone from address book
  **-> tryNormalizePhoneNumberToE164(raw, SIM/locale country)
         -> invalid: drop
         -> valid: PhoneNumber(E.164)**
  -> msisdnSanitizer
  -> hash
  -> lookup

## Examples
| Input | callerIsoCode | return E.164 | msisdn |
|---|---|---|---|
| `0612345678` | `FR` | `+33612345678` | `33612345678` |
| `06 12 34 56 78` | `FR` | `+33612345678` | `33612345678` |
| `2125556789` | `US` | `+12125556789` | `12125556789` |
| `0033612345678` | `FR` | `+33612345678` | `33612345678` |
| `+33 (0) 6 12 34 56 78` | `FR` | `+33612345678` | `33612345678` |
| `+33612345678` | `null` | `+33612345678` | `33612345678` |
| `0612345678` | `null` | `null` | `n/a` |
| `not-a-number` | `FR` | `null` | `n/a` |
| `+7 912 345-67-89` | `FR` | `+79123456789` | `79123456789` |
| `+7 (495) 123-45-67` | `FR` | `+74951234567` | `74951234567` |

## Impact description
- Improves reliability of local contact lookup by ensuring phone inputs are normalized and valid before hash generation.
- Reduces false negatives caused by formatting noise in phonebook entries.
- Does not change Matrix lookup protocol shape (same request fields/algorithm flow): it improves input quality.

## Test recommendations
- Run:
  - `flutter test test/data/datasource_impl/contact/phonebook_contact_datasource_impl_test.dart`
  - `flutter test test/utils/phone_number_normalizer_test.dart`
  - `flutter test test/domain/usecase/contacts/twake_look_up_phonebook_contact_interactor_test.dart`
- Manual checks recommended:
  - local numbers without `+` with SIM country available
  - local numbers without `+` with SIM unavailable but locale fallback active
  - malformed numbers / extension-heavy values are not sent to lookup hashing
- Additional caution:
  - `SimCountryProvider` currently caches first resolution result (including `null`). If first call fails transiently, fallback behavior may persist for session.
 
## Demos
- N/A (backend/domain flow change, no UI behavior demo required).

## Summary by CodeRabbit

* **New Features**
  * Phone numbers are normalized to international E.164 for consistent display and deduplication.
  * Device/SIM-based country detection is used to improve normalization of local numbers (with safe web fallback).

* **Bug Fixes**
  * Contacts with unresolvable or invalid phone numbers are excluded to reduce errors and duplicates.

* **Documentation**
  * Added a carrier-usage privacy notice explaining use of carrier info to infer a default country code.

### CODE REVIEW

File: lib/data/datasource/contact/sim_country/sim_country_provider_impl_io.dart [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import 'dart:io';
```
[the remark this answers is not in the document]
Comment (jbrnzl) [reply lib/data/datasource/contact/sim_country/sim_country_provider_impl_io.dart:1]: Web builds will fail to compile: this is false. The CI is green also. But anyway you are right in the theory so I updated the code and seperated things to make it less confuse in the future.

File: lib/data/datasource/contact/sim_country_provider.dart [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  @override
+  Future<String?> getCountryCode() async {
+    if (_fetched) return _cache;
+    _fetched = true;
+    _cache = await _resolve();
+    return _cache;
+  }
+
+  bool _isValidSimCode(String? code) =>
+      code != null && code.isNotEmpty && code != '--';
```
[the remark this answers is not in the document]
Comment (jbrnzl) [reply lib/data/datasource/contact/sim_country_provider.dart]: Adressed

File: pubspec.yaml [hunk 3]
```diff
@@ -194,6 +194,8 @@ dependencies:
   super_clipboard: 0.9.1
   crypto: ^3.0.3
   flutter_contacts: ^2.0.0-beta.2
+  phone_numbers_parser: ^9.0.21
+  sim_country_code_plus: ^0.1.4
```
Comment (nqhhdev) [inline pubspec.yaml:198]: Hi @jbrnzl, it looks like the pubspec.lock file is missing from your PR.

Comment (jbrnzl) [reply pubspec.yaml:198]: Hi @nqhhdev, I added pubspec.lock 👍🏻

File: lib/data/datasource/contact/sim_country/sim_country_provider_impl_io.dart [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+class SimCountryProviderImpl implements SimCountryProvider {
+  Future<String?>? _countryCodeFuture;
+
+  @override
+  Future<String?> getCountryCode() => _countryCodeFuture ??= _resolve();
+
+  // Only accept strict ISO 3166-1 alpha-2 codes (exactly two ASCII letters).
+  bool _isValidSimCode(String? code) =>
+      code != null && RegExp(r'^[A-Za-z]{2}$').hasMatch(code);
```
[the remark this answers is not in the document]
Comment (jbrnzl) [reply lib/data/datasource/contact/sim_country/sim_country_provider_impl_io.dart]: Ok I understand you point. I am thinking about modifying like this: code != null && pnp.isoCodeConversionMap.containsKey(code.toUpperCase()). What do you think ? It will be simpler no ? That way if we have 'ZZ' validation will fail and fallback on system locale

**Review by nqhhdev — APPROVED** [review]

**Review by pm-McFly — APPROVED** [review]
